core/src/layers/madsim.rs

```diff
@@ -0,0 +1,316 @@
+use crate::raw::oio::Entry;
+use crate::raw::{oio, Accessor, Layer, LayeredAccessor, RpList, RpRead, RpScan, RpWrite};
+use crate::{EntryMode, Metadata};
+use async_trait::async_trait;
+use bytes::Bytes;
+use madsim::plugin::simulator;
+use madsim::plugin::Simulator;
+use madsim::rand::GlobalRng;
+use madsim::runtime::Handle;
+use madsim::task::NodeId;
+use madsim::time::TimeHandle;
+use madsim::Config;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::SeekFrom;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+#[derive(Debug, Copy, Clone)]
+pub struct MadsimLayer {
+    service_sim_node: NodeId,
+}
+
+impl MadsimLayer {
+    pub fn new() -> Self {
+        let service_sim_node = Handle::current().create_node().build().id();
+        Self { service_sim_node }
+    }
+}
+
+impl<A: Accessor> Layer<A> for MadsimLayer {
+    type LayeredAccessor = MadsimAccessor;
+
+    fn layer(&self, _inner: A) -> Self::LayeredAccessor {
+        let handle = simulator::<ServiceSim>().get_node(self.service_sim_node);
+        MadsimAccessor { handle }
+    }
+}
+
+#[derive(Debug)]
+pub struct MadsimAccessor {
+    handle: ServiceNodeHandle,
+}
+
+#[async_trait]
+impl LayeredAccessor for MadsimAccessor {
+    type Inner = ();
+    type Reader = MadsimReader;
+    type BlockingReader = MadsimReader;
+    type Writer = MadsimWriter;
+    type BlockingWriter = MadsimWriter;
+    type Pager = MadsimPager;
+    type BlockingPager = MadsimPager;
+
+    fn inner(&self) -> &Self::Inner {
+        &()
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> crate::Result<(RpRead, Self::Reader)> {
+        let handle = self.handle.inner.lock().unwrap();
+        let length = handle
+            .get(path)
+            .map(|(_, data)| data.len())
+            .unwrap_or_default();
+        Ok((
+            RpRead::new(length as u64),
+            MadsimReader {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> crate::Result<(RpWrite, Self::Writer)> {
+        Ok((
+            RpWrite::new(),
+            MadsimWriter {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn list(&self, path: &str, args: OpList) -> crate::Result<(RpList, Self::Pager)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    async fn scan(&self, path: &str, args: OpScan) -> crate::Result<(RpScan, Self::Pager)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn blocking_read(
+        &self,
+        path: &str,
+        args: OpRead,
+    ) -> crate::Result<(RpRead, Self::BlockingReader)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
```

Comment:
Generally we don't support running a blocking function in the simulation, as it will block the only execution thread. In other words, all operations which can not complete immediately should be defined as async functions.
So these blocking-prefixed functions should be either unsupported at all, or be considered happen at an instant.